### PR TITLE
feat: Validate messages in the Snuba eventstream backend

### PR DESCRIPTION
**Pull request by lynnagara**
The Snuba backend is only used in dev and CI, no validation has been added in the Kafka backend... yet

Currently this only validates messages on the "events" topic. "generic-events" and "transactions" should be tackled as follow ups.

### CODE REVIEW

File: src/sentry/data/samples/empty-exception.json [hunk 1]
```diff
@@ -1,38 +1,29 @@
...
-        "setuptools": "40.0.0",
-        "urllib3": "1.23",
-        "wheel": "0.31.1",
-        "wsgiref": "0.1.2"
-    },
-    "platform": "python",
-    "server_name": "mcbk.local",
-    "exception": {
-        "values": [
-            null
```
Comment (untitaker) [inline src/sentry/data/samples/empty-exception.json:17]: this can happen though..

File: src/sentry/eventstream/base.py [hunk 2]
```diff
@@ -187,7 +187,7 @@ def tombstone_events_unsafe(
         self,
         project_id: int,
         event_ids: Sequence[str],
-        old_primary_hash: Union[str, bool] = False,
```
Comment (untitaker) [inline src/sentry/eventstream/base.py:190]: lol. I ran git blame and wondered who tf decided that `old_primary_hash` should be a boolean... turns out I decided that, and you just added the type annotation. `None` is indeed more correct.

File: src/sentry/eventstream/snuba.py [hunk 3]
```diff
@@ -32,6 +32,8 @@
 if TYPE_CHECKING:
     from sentry.eventstore.models import Event
 
+DATETIME_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"
```
Comment (untitaker) [inline src/sentry/eventstream/snuba.py:35]: i am sure that this constant is already defined in `sentry.utils.json` somewhere?

File: src/sentry/eventstream/snuba.py [hunk 4]
```diff
@@ -419,11 +421,30 @@ def _send(
...
+
+        try:
+            codec = sentry_kafka_schemas.get_codec(
+                topic={
+                    "events": "events",
+                    # "transactions": "transactions",
+                    "search_issues": "generic-events",
+                }[entity]
+            )
+        except Exception:
```
Comment (untitaker) [inline src/sentry/eventstream/snuba.py:435]: can you do `(SchemaNotFound, KeyError)` here?

**Review by untitaker — APPROVED** [review]
I wonder if we should add support for custom json serializers in codecs, so we can support things like datetime being automatically serialized to the right string format

File: src/sentry/eventstream/snuba.py (hunk 3 above)
Comment (lynnagara) [reply src/sentry/eventstream/snuba.py:35]: There was only a copy in synchronized.py. I moved it here since it wasn't used there.

Comment (untitaker) [reply src/sentry/eventstream/snuba.py:35]: i mean, can you use this instead? https://github.com/getsentry/sentry/blob/236822aa884450b3a2e6ca4a0b2aa71fb2b20cdf/src/sentry/utils/json.py#L26-L27

this is the function we serialize datetimes with today

Comment (lynnagara) [reply src/sentry/eventstream/snuba.py:35]: Oh I missed that one, will do

**Review by untitaker — APPROVED** [review]

**Pull request merged**